src/include/cpu/rpp_cpu_filter.hpp

```diff
@@ -0,0 +1,568 @@
+    pxTemp[2] = _mm_shuffle_epi8(_mm_blend_epi16(px128[1], px128[2], 1), xmm_pxMaskRotate0To1);    // blend with mask [0000 0001] and shuffle - [R04|G04|B04|R05|G05|B05|R06|G06]
+    pxTemp[3] = _mm_shuffle_epi8(_mm_blend_epi16(px128[1], px128[2], 15), xmm_pxMaskRotate0To7);   // blend with mask [0000 1111] and shuffle - [R05|G05|B05|R06|G06|B06|R07|G07]
+    pxTemp[4] = _mm_shuffle_epi8(_mm_blend_epi16(px128[1], px128[2], 127), xmm_pxMaskRotate0To13); // blend with mask [0111 1111] and shuffle - [R06|G06|B06|R07|G07|B07|R08|G08]
+    pxTemp[5] = _mm_shuffle_epi8(_mm_blend_epi16(px128[2], px128[3], 3), xmm_pxMaskRotate0To3);    // blend with mask [0000 0011] and shuffle - [R07|G07|B07|R08|G08|B08|R09|G09]
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[0]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[1]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[2]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[3]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[4]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[5]);
+}
+
+inline void blend_shuffle_add_9x9_pln_host(__m128i *px128)
+{
+    /* px128[0] - [X01|X02|X03|X04|X05|X06|X07|X08]
+       px128[1] - [X09|X10|X11|X12|X13|X14|X15|X16] */
+    __m128i pxTemp[7];
+    pxTemp[0] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 1), xmm_pxMaskRotate0To1);    // blend with mask [0000 0001] and shuffle - [X02|X03|X04|X05|X06|X07|X08|X09]
+    pxTemp[1] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 3), xmm_pxMaskRotate0To3);    // blend with mask [0000 0011] and shuffle - [X03|X04|X05|X06|X07|X08|X09|X10]
+    pxTemp[2] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 7), xmm_pxMaskRotate0To5);    // blend with mask [0000 0111] and shuffle - [X04|X05|X06|X07|X08|X09|X10|X11]
+    pxTemp[3] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 15), xmm_pxMaskRotate0To7);   // blend with mask [0000 1111] and shuffle - [X05|X06|X07|X08|X09|X10|X11|X12]
+    pxTemp[4] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 31), xmm_pxMaskRotate0To9);   // blend with mask [0001 1111] and shuffle - [X06|X07|X08|X09|X10|X11|X12|X13]
+    pxTemp[5] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 63), xmm_pxMaskRotate0To11);  // blend with mask [0011 1111] and shuffle - [X07|X08|X09|X10|X11|X12|X13|X14]
+    pxTemp[6] = _mm_shuffle_epi8(_mm_blend_epi16(px128[0], px128[1], 127), xmm_pxMaskRotate0To13); // blend with mask [0111 1111] and shuffle - [X08|X09|X10|X11|X12|X13|X14|X15]
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[0]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[1]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[2]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[3]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[4]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[5]);
+    px128[0] = _mm_add_epi16(px128[0], pxTemp[6]);
+    px128[0] = _mm_add_epi16(px128[0], px128[1]);
+}
+inline void blend_shuffle_add_9x9_pkd_host(__m128i *px128)
```

Comment:
In the set of blend-shuffle functions here, there are sets of two functions for each type (one for pln and one for pkd).
Can we combine to have 1?

If we pass the entire __m128i mask array into the function, and an argument that separates pln from pkd, it could possibly be just 1 function.
The argument that separates pln from pkd could be a pointer to a const permutation index array - pointer to [0,1,2,3,4,5,6] vs [2,5,0,3,6,1,4] and you can address inside the function as xmm_pxMaskRotate[permuteArr[0/1/2/3/4/5/6]]?

Reply:
Done